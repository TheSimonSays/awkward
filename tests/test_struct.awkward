struct Product {
    name;
    price;
    quantity;
};

impl Product {
    fn total_value() {
        return self.price * self.quantity;
    }

    fn restock(amount) {
        self.quantity = self.quantity + amount;
        return "Restocked " + self.name + " by " + amount + " units";
    }
}

fn sell_product(product, amount) {
    if (product.quantity >= amount) {
        product.quantity = product.quantity - amount;
        return product.price * amount;
    } else {
        return "Error Not enough stock for " + product.name;
    }
}


let phone = new Product{name="Apple", price=1000, quantity=10};
print("Initial stock " + phone);
print(phone.restock(5));
print("After restock " + phone);
print("Total value " + phone.total_value());


let sales = 0;
let i = 0;

while (i < 3) {
    let sale = sell_product(phone, 1);
    print("Sale " + (i + 1) + " " + sale);
    sales = sales + sale;
    i = i + 1;
}

print("Total sales " + sales);

print("Final stock " + phone);

let phone1 = new Product{name="Samsung", price=100, quantity=1};

print(phone == phone1);

struct Storage {
    data;
};

let some_storage = new Storage{data=[]};

print(some_storage.data);

some_storage.data.extend([1, 2, 3]);

print(some_storage.data);

for (let i in some_storage) {
    print(i, " ", some_storage[i]);
    if (type(some_storage[i]) == "array") {
        print("struct value is array");
        for (let j in some_storage[i]) {
            print("struct data value by ", j " is ", some_storage[i][j]);
        }
    }
}
