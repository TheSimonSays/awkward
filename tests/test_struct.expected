Initial stock {name: Apple, price: 1000, quantity: 10}
Restocked Apple by 5 units
After restock {name: Apple, price: 1000, quantity: 15}
Total value 15000
Sale 1 1000
Sale 2 1000
Sale 3 1000
Total sales 3000
Final stock {name: Apple, price: 1000, quantity: 12}
false
[]
[1, 2, 3]
data [1, 2, 3]
struct value is array
struct data value by 0 is 1
struct data value by 1 is 2
struct data value by 2 is 3
